src/slice/group_by.rs

```diff
@@ -0,0 +1,194 @@
+use crate::iter::plumbing::*;
+use crate::iter::*;
+
+fn find_index<T, F>(xs: &[T], pred: &F) -> Option<usize>
+where
+    F: Fn(&T, &T) -> bool,
+{
+    let n = (xs.len() / 2).saturating_sub(1);
+
+    for m in (1..((n / 2) + 1)).map(|x| 2 * x) {
+        let start = n.saturating_sub(m);
+        let end = std::cmp::min(n + m, xs.len());
+        let fsts = &xs[start..end];
+        let (_, snds) = fsts.split_first()?;
+        match fsts.iter().zip(snds).position(|(x , y)| !pred(x, y)) {
```

Comment:
If I understand correctly, these nested loops are performing a widening search, right? With `n` as the midpoint, it's basically scanning `mid-2..mid+2`, `mid-4..mid+4`, `mid-6..mid+6`, etc.

If so, that's O(n^2) as it will keep re-testing the middle portion that the previous outer loop already tried.

Reply:
Yes I believe you're correct. Or at least that's what I was trying to accomplish. It was the most obviously correct way for me. I had an optimisation planned where it doesn't rescan the center but haven't yet worked it out in a way that preserves correctness.